### PR TITLE
[CAS-419] Attachment gallery

### PR DESCRIPTION
**Pull request by leandroBorgesFerreira**
[CAS-419](https://stream-io.atlassian.net/browse/CAS-419)

### Description

This PR creates the custom view for an image gallery for all attachments. The view is added to the Components Browser.

The result: 
Nov-27-2020 01-39-34

**Note: After the reviews, I added some commits that made relevants changes in the PR, please consider it when reviewing it again.**

### Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [x] Changelog updated with client-facing changes
- [ ] New code is covered by unit tests. **Not necessary.**
- [x] Comparison screenshots added for visual changes.
- [x] Reviewers added

giphy

### CODE REVIEW

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/AttachmentGallery.kt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import io.getstream.chat.android.ui.databinding.StreamAttachmentGalleryBinding
+
+public class AttachmentGallery : ConstraintLayout {
+
+    private val binding = StreamAttachmentGalleryBinding.inflate(LayoutInflater.from(context), this, true)
+
+    private var countText: String = "%s - %s"
+
+    public constructor(context: Context) : super(context) {
+        init(null)
```
Comment (ogkuzmin) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/AttachmentGallery.kt:21]: Why do you invoke it if it do nothing?

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/ImageSlidePageFragment.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.fragment.app.Fragment
+import com.getstream.sdk.chat.ImageLoader
+import io.getstream.chat.android.ui.databinding.StreamItemImageGalleryBinding
+
+public class ImageSlidePageFragment : Fragment() {
+
+    public var image: String? = null
```
Comment (ogkuzmin) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/ImageSlidePageFragment.kt:13]: nit: `imageUrl` describes it more accurate then just `image`

**Review by ogkuzmin — APPROVED** [review]

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/AttachmentGallery.kt (hunk 1 above)
Comment (leandroBorgesFerreira) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/AttachmentGallery.kt:21]: Just a mistake, I took it from draft too early. Let me fix this.

**Review by ogkuzmin — DISMISSED** [review]

**Review by samiuelson — APPROVED** [review]

File: stream-chat-android-ui-components-sample/src/main/res/layout/fragment_user_login.xml [hunk 3]
```diff
@@ -85,4 +85,13 @@
         app:layout_constraintStart_toStartOf="parent"
         tools:text="@string/sdk_version_template"
         />
+
+    <io.getstream.chat.android.ui.images.AttachmentGallery
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components-sample/src/main/res/layout/fragment_user_login.xml:89]: Why are we adding it to user login fragment?

Comment (leandroBorgesFerreira) [reply stream-chat-android-ui-components-sample/src/main/res/layout/fragment_user_login.xml:89]: Doesn't make any sense. I'll remove this in my next commit.

**Comment by adasiewiczr** [comment]
It would be great to add it to our `ComponentBrowser` views

**Comment by bychkovdmitry** [comment]
Looks good, but doesn't run on my device. The sample app just shows a black screen.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/AttachmentSlidePagerAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+internal class AttachmentSlidePagerAdapter(
+    fa: FragmentActivity,
+    private val imageList: List<String>
+) : FragmentStateAdapter(fa) {
+
+    override fun getItemCount(): Int = imageList.size
+
+    override fun createFragment(position: Int): Fragment {
+        return ImageSlidePageFragment().apply {
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/AttachmentSlidePagerAdapter.kt:15]: Do we support device rotation? This won't work on configuration change

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/ImageSlidePageFragment.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        return StreamItemImageGalleryBinding.inflate(LayoutInflater.from(context)).apply {
+            binding = this
+        }.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        imageUrl?.let {
+            ImageLoader.run {
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/ImageSlidePageFragment.kt:31]: IMO, import is better
```
imageUrl?.let {
    binding.ivImageItem.load(it)
}
```

File: stream-chat-android-ui-components/src/main/res/layout/stream_attachment_gallery.xml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    >
+
+    <androidx.viewpager2.widget.ViewPager2
+        android:id="@+id/vp_attachmentGallery"
+        android:layout_width="match_parent"
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components/src/main/res/layout/stream_attachment_gallery.xml:11]: Not a big deal, but I had some subtle bugs when using `match_parent`. `0dp` is better.

File: stream-chat-android-ui-components/src/main/res/layout/stream_attachment_gallery.xml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        android:layout_height="0dp"
+        android:layout_marginStart="4dp"
+        android:layout_marginEnd="4dp"
+        android:layout_marginBottom="16dp"
+        app:layout_constraintBottom_toTopOf="@id/tv_photoCount"
+        app:layout_constraintTop_toTopOf="parent"
+        />
+
+    <TextView
+        android:id="@+id/tv_photoCount"
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components/src/main/res/layout/stream_attachment_gallery.xml:21]: We need to agree on the naming and align everything in `ui-components` and `ui-components-sample` modules. IMO, the combination of camel case and snake case looks not so good.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/AttachmentSlidePagerAdapter.kt (hunk 4 above)
Comment (bychkovdmitry) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/AttachmentSlidePagerAdapter.kt:15]: Even though the adapter is recreated on rotation, the fragments are restored from bundle

File: stream-chat-android-ui-components/src/main/res/layout/stream_item_image_gallery.xml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<FrameLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <com.github.chrisbanes.photoview.PhotoView
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components/src/main/res/layout/stream_item_image_gallery.xml:7]: Expecting something else in this layout? Maybe better to use PhotoView as a root layout?

File: stream-chat-android-ui-components/src/main/res/layout/stream_item_image_gallery.xml [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+        android:id="@+id/iv_ImageItem"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:src="@drawable/stream_ic_attach"
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components/src/main/res/layout/stream_item_image_gallery.xml:11]: I think placeholders should be a part of coil loading logic

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/ImageSlidePageFragment.kt (hunk 5 above)
Comment (bychkovdmitry) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/ImageSlidePageFragment.kt:31]: I think it's fine to pass null here `binding.ivImageItem.load(nullable)`

File: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/component_browser/attachments/gallery/ComponentBrowserAttachmentGalleryFragment.kt [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.Toast
+import androidx.fragment.app.Fragment
+import androidx.navigation.fragment.navArgs
+import io.getstream.chat.ui.sample.databinding.FragmentComponentBrowserAttachmentGalleryBinding
+
+class ComponentBrowserAttachmentGalleryFragment : Fragment() {
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/component_browser/attachments/gallery/ComponentBrowserAttachmentGalleryFragment.kt:12]: Shouldn't this be a part of ui-components library?

**Review by bychkovdmitry — CHANGES_REQUESTED** [review]

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/ImageSlidePageFragment.kt (hunk 5 above)
Comment (leandroBorgesFerreira) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/ImageSlidePageFragment.kt:31]: No problem to pass a nullable. But I need the `run` to change the scope to I can use the extension functions of ImageLoader.

File: stream-chat-android-ui-components/src/main/res/layout/stream_attachment_gallery.xml (hunk 7 above)
Comment (leandroBorgesFerreira) [reply stream-chat-android-ui-components/src/main/res/layout/stream_attachment_gallery.xml:21]: 🤔  let's use photoCount then.

**Review by bychkovdmitry — APPROVED** [review]

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/ImageSlidePageFragment.kt (hunk 5 above)
Comment (zsmb13) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/ImageSlidePageFragment.kt]: You can import it with `import com.getstream.sdk.chat.ImageLoader.load` to make it work cleanly.

Though it was probably a bad design choice to make it a member extension in the first place, member extensions are weird. We could move these into top level extensions instead, if we want to.

Comment (leandroBorgesFerreira) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/images/ImageSlidePageFragment.kt]: done

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,4 +1,4 @@
-# To be released:
```
Comment (JcMinarro) [inline CHANGELOG.md:1]: This text should be keeped

File: CHANGELOG.md [hunk 12]
```diff
@@ -34,6 +34,7 @@
 ## stream-chat-android-ui-components
 - Add MessagesHeaderView
 - New artifact name: `io.getstream:stream-chat-android-ui-components:STREAM_VERSION`
+- Adding AttachmentGallery to handle attachments
```
Comment (JcMinarro) [inline CHANGELOG.md:37]: It seems that you added changelog on wrong place

Comment (zsmb13) [reply CHANGELOG.md:37]: We could avoid merge/rebase problems in the changelog if the "to be released" parts were kept in a separate file, and moved to the actual changelog when the release is created. What do you think, @JcMinarro?

**Review by bychkovdmitry — DISMISSED** [review]

File: CHANGELOG.md (hunk 12 above)
Comment (leandroBorgesFerreira) [reply CHANGELOG.md:37]: Damn... Sorry merge problem.

**Review by zsmb13 — APPROVED** [review]

File: CHANGELOG.md (hunk 12 above)
Comment (JcMinarro) [reply CHANGELOG.md]: I don't really see a problem with this behavior. Having it into a separated file could be the same, add some changes that are previously added in other version or totally remove the new changes added to the "to be released" file.
We can discuss if we would prefer a new file or any other alternative

**Review by zsmb13 — APPROVED** [review]

**Pull request merged**